### PR TITLE
Logger: Add basic userspace support

### LINKED ISSUE #11743: logging: add user mode access
We'd like to allow user threads to use the logging subsystem, at the moment only supervisor threads may log events.

How to do this will require some thought. We want to minimize latency and also ensure that garbage passed in will not present any opportunities to crash or compromise the kernel. We'd also like to use the same logging APIs in user mode, and not have a parallel set.

Further comments for analysis and brainstorming, I don't have a concrete implementation proposal yet.

### PR DESCRIPTION
**Pull request by pizi-nordic**
This PR adds basic userspace support to the logging subsystem.
With this change, the following API could be called from user mode:        
    - LOG_*()                                                                  
    - LOG_INST_*(),                                                            
    - LOG_HEXDUMP_*(),                                                         
    - LOG_HEXDUMP_INST_*(),                                                    
    - LOG_PANIC(), LOG_PROCESS(),                                              
    - log_printk(), log_generic(), log_buffrered_cnt(),                        
    - log_filter_set(NULL, ...)                                                

With userspace disabled, the logger behaviour and performance is not affected. With userspace enabled, the calls from kernelspace have an additional overhead introduced by _is_user_context().        

The logger behaviour changes when it is called from the user context.
All strings logged using LOG_*() and LOG_INST_*() API from userspace
are rendered in place for security reasons and then placed in log_strdup()
memory pool, which should be large enough to hold bursts of log messages.

Fixes #11743

### CODE REVIEW

File: include/logging/log_core.h [hunk 1]
```diff
@@ -162,9 +163,11 @@ extern "C" {
 
 #define Z_LOG_INTERNAL_X(N, ...)  UTIL_CAT(_LOG_INTERNAL_, N)(__VA_ARGS__)
 
-#define __LOG_INTERNAL(_src_level, ...)					 \
+#define __LOG_INTERNAL(is_user_context, _src_level, ...)		 \
 	do {								 \
-		if (IS_ENABLED(CONFIG_LOG_IMMEDIATE)) {		 \
+		if (is_user_context) {					 \
```
Comment (carlescufi) [inline include/logging/log_core.h]: why is this not enclosed in a `#if CONFIG_USERSPACE`? won't having the `if` statement add code even when userspace is disabled?

File: include/logging/log_core.h [hunk 2]
```diff
@@ -212,19 +215,23 @@ extern "C" {
 /******************************************************************************/
 #define __LOG(_level, _id, _filter, ...)				    \
 	do {								    \
+		bool is_user_context = _is_user_context();		    \
```
Comment (carlescufi) [inline include/logging/log_core.h]: see comment above

File: include/logging/log_core.h [hunk 3]
```diff
@@ -254,15 +261,21 @@ extern "C" {
 /******************************************************************************/
 #define __LOG_HEXDUMP(_level, _id, _filter, _data, _length, _str)	      \
 	do {								      \
+		bool is_user_context = _is_user_context();		      \
```
Comment (carlescufi) [inline include/logging/log_core.h]: see comment above

File: subsys/logging/log_core.c [hunk 4]
```diff
@@ -324,27 +325,37 @@ int log_printk(const char *fmt, va_list ap)
...
+			u32_t value;
+		} src_level_union = {
+			{
+				.level = LOG_LEVEL_INTERNAL_RAW_STRING
+			}
 		};
 
-		if (IS_ENABLED(CONFIG_LOG_IMMEDIATE)) {
-			log_generic(src_level, fmt, ap);
+		if (_is_user_context()) {
```
Comment (carlescufi) [inline subsys/logging/log_core.c]: Same, no need for `#ifdef`?

File: subsys/logging/log_core.c [hunk 5]
```diff
@@ -375,7 +386,9 @@ static u32_t count_args(const char *fmt)
 
 void log_generic(struct log_msg_ids src_level, const char *fmt, va_list ap)
 {
-	if (IS_ENABLED(CONFIG_LOG_IMMEDIATE) &&
+	if (_is_user_context()) {
```
Comment (carlescufi) [inline subsys/logging/log_core.c]: same as above.

I assume you rely on the compiler to automatically remove these, just wondering if you've checked

**Review by carlescufi — COMMENTED** [review]
Looks good, just wondering about compiler optimizations

File: include/logging/log_core.h (hunk 1 above)
Comment (pizi-nordic) [reply include/logging/log_core.h]: This is a macro definition, where I cannot use #ifdefs. So if I would like to use this classic method, I would have to several definitions of the macros here.

As you noted, I a relying on the fact, that value returned by the `_is_user_context()` is known at compile time when userspace is disabled:
```
/**
 * Indicate whether the CPU is currently in user mode
 *
 * @return true if the CPU is currently running with user permissions
 */
static inline bool _is_user_context(void)
{
#ifdef CONFIG_USERSPACE
       return z_arch_is_user_context();
#else
       return false;
#endif
}
```
Unfortunatelly, the dead code is not removed when you are compiling with `CONFIG_NO_OPTIMIZATIONS=y`, but I assume that additional overhead will be not a problem in such case.

File: include/logging/log_core.h (hunk 2 above)
Comment (pizi-nordic) [reply include/logging/log_core.h]: Addressed above.

File: include/logging/log_core.h (hunk 3 above)
Comment (pizi-nordic) [reply include/logging/log_core.h]: Addressed above.

File: subsys/logging/log_core.c (hunk 4 above)
Comment (pizi-nordic) [reply subsys/logging/log_core.c]: Addressed above.

File: subsys/logging/log_core.c (hunk 5 above)
Comment (pizi-nordic) [reply subsys/logging/log_core.c]: Addressed above.

**Review by carlescufi — APPROVED** [review]
@pizi-nordic thanks for the explanations, LGTM

**Review by nashif — APPROVED** [review]

File: include/syscall.h [hunk 6]
```diff
@@ -122,16 +122,6 @@ typedef u32_t (*_k_syscall_handler_t)(u32_t arg1, u32_t arg2, u32_t arg3,
  */
 static inline bool z_arch_is_user_context(void);
```
Comment (ioannisg) [inline include/syscall.h]: Could you move this declaration below, as well? So it is close to the _is_user_context definition

Comment (pizi-nordic) [reply include/syscall.h]: Done

**Pull request merged**